File: ios/HealthTracker/Services/MealPlanLoader.swift
import Foundation

// MARK: - JSON Models for Decoding
struct MealPlanJSON: Codable {
    let dietType: String
    let displayName: String
    let description: String
    let benefits: [String]
    let restrictions: [String]
    let meals: [String: MealJSON]
    let weeks: [WeekJSON]
}

struct MealJSON: Codable {
    let name: String
    let description: String
    let calories: Int
    let protein: Double
    let carbs: Double
    let fat: Double
    let fiber: Double
    let prepTime: Int
    let cookTime: Int
    let ingredients: [String]
    let instructions: [String]
    let tags: [String]
}

struct WeekJSON: Codable {
    let weekNumber: Int
    let days: [DayScheduleJSON]
}

struct DayScheduleJSON: Codable {
    let day: String
    let breakfast: String
    let lunch: String
    let dinner: String
    let snack: String
}

// MARK: - Meal Plan Loader Service
class MealPlanLoader {
    static let shared = MealPlanLoader()

    private var cache: [String: MealPlanType] = [:]

    private let dietFiles: [(id: String, file: String)] = [
        ("mediterranean", "mediterranean"),
        ("keto", "keto"),
        ("high_protein", "high_protein"),
        ("balanced", "balanced"),
        ("low_carb", "low_carb"),
        ("paleo", "paleo"),
        ("whole30", "whole30"),
        ("vegan", "vegan")
    ]

    private init() {}

    // MARK: - Load All Meal Plans
    func loadAllMealPlans() -> [MealPlanType] {
        return dietFiles.compactMap { loadMealPlan(for: $0.id, file: $0.file) }
    }

    // MARK: - Load Single Meal Plan
    func loadMealPlan(for dietId: String, file: String? = nil) -> MealPlanType? {
        if let cached = cache[dietId] {
            return cached
        }

        let fileName = file ?? dietId
        guard let url = Bundle.main.url(forResource: fileName, withExtension: "json", subdirectory: "MealPlans"),
              let data = try? Data(contentsOf: url) else {
            // Fallback: try without subdirectory (flat bundle)
            guard let url = Bundle.main.url(forResource: fileName, withExtension: "json"),
                  let data = try? Data(contentsOf: url) else {
                return nil
            }
            return decodeMealPlan(from: data, dietId: dietId)
        }

        return decodeMealPlan(from: data, dietId: dietId)
    }

    private func decodeMealPlan(from data: Data, dietId: String) -> MealPlanType? {
        guard let json = try? JSONDecoder().decode(MealPlanJSON.self, from: data) else {
            return nil
        }

        let plan = convertToMealPlanType(from: json)
        cache[dietId] = plan
        return plan
    }

    // MARK: - Convert JSON to Model Structs
    func convertToMealPlanType(from json: MealPlanJSON) -> MealPlanType {
        let weeklyPlans = json.weeks.map { weekJSON -> WeeklyMealPlan in
            let days = weekJSON.days.map { dayJSON -> DailyMealPlan in
                let dayId = "\(json.dietType)-w\(weekJSON.weekNumber)-\(dayJSON.day.lowercased())"

                let breakfast = resolveMeal(id: dayJSON.breakfast, from: json.meals, prefix: json.dietType)
                let lunch = resolveMeal(id: dayJSON.lunch, from: json.meals, prefix: json.dietType)
                let dinner = resolveMeal(id: dayJSON.dinner, from: json.meals, prefix: json.dietType)
                let snack = resolveMeal(id: dayJSON.snack, from: json.meals, prefix: json.dietType)

                return DailyMealPlan(
                    id: dayId,
                    dayName: dayJSON.day,
                    breakfast: breakfast,
                    lunch: lunch,
                    dinner: dinner,
                    snacks: [snack]
                )
            }

            return WeeklyMealPlan(
                id: "\(json.dietType)-week-\(weekJSON.weekNumber)",
                weekNumber: weekJSON.weekNumber,
                days: days
            )
        }

        return MealPlanType(
            id: json.dietType,
            name: json.displayName,
            description: json.description,
            benefits: json.benefits,
            restrictions: json.restrictions,
            monthlyPlans: weeklyPlans
        )
    }

    private func resolveMeal(id: String, from catalog: [String: MealJSON], prefix: String) -> Meal {
        guard let mealJSON = catalog[id] else {
            // Fallback for missing meal references
            return Meal(
                id: id,
                name: "Meal",
                description: "",
                calories: 0,
                protein: 0,
                carbs: 0,
                fat: 0,
                fiber: 0,
                prepTime: 0,
                cookTime: 0,
                ingredients: [],
                instructions: [],
                tags: []
            )
        }

        return Meal(
            id: id,
            name: mealJSON.name,
            description: mealJSON.description,
            calories: mealJSON.calories,
            protein: mealJSON.protein,
            carbs: mealJSON.carbs,
            fat: mealJSON.fat,
            fiber: mealJSON.fiber,
            prepTime: mealJSON.prepTime,
            cookTime: mealJSON.cookTime,
            ingredients: mealJSON.ingredients,
            instructions: mealJSON.instructions,
            tags: mealJSON.tags
        )
    }
}

File: android/app/src/main/java/com/mochasmindlab/mlhealth/services/MealPlanLoader.kt
package com.mochasmindlab.mlhealth.services

import android.content.Context
import com.mochasmindlab.mlhealth.data.models.DailyMealPlan
import com.mochasmindlab.mlhealth.data.models.DietPlan
import com.mochasmindlab.mlhealth.data.models.MealPlanJson
import com.mochasmindlab.mlhealth.data.models.PlanRecipe
import com.mochasmindlab.mlhealth.data.models.PlanRecipeJson
import com.mochasmindlab.mlhealth.data.models.WeeklyMealPlan
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Loads bundled meal plan JSON files from `assets/MealPlans/`.
 * Mirrors iOS MealPlanLoader.swift.
 */
@Singleton
class MealPlanLoader @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val cache = ConcurrentHashMap<String, DietPlan>()

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    suspend fun loadAll(): List<DietPlan> = withContext(Dispatchers.IO) {
        DIET_FILES.mapNotNull { (id, file) -> loadDietPlan(id, file) }
    }

    suspend fun loadDietPlan(dietId: String, file: String? = null): DietPlan? = withContext(Dispatchers.IO) {
        cache[dietId]?.let { return@withContext it }

        val fileName = file ?: dietId
        val plan = runCatching {
            val text = context.assets.open("MealPlans/$fileName.json").use { input ->
                input.bufferedReader().readText()
            }
            val parsed = json.decodeFromString(MealPlanJson.serializer(), text)
            convert(parsed)
        }.getOrNull()

        plan?.also { cache[dietId] = it }
    }

    private fun convert(parsed: MealPlanJson): DietPlan {
        val recipeMap = parsed.meals.mapValues { (id, recipe) -> recipe.toPlanRecipe(id) }

        val weeks = parsed.weeks.map { week ->
            WeeklyMealPlan(
                weekNumber = week.weekNumber,
                days = week.days.map { day ->
                    DailyMealPlan(
                        dayName = day.day,
                        breakfast = recipeMap[day.breakfast] ?: missingPlanRecipe(day.breakfast),
                        lunch = recipeMap[day.lunch] ?: missingPlanRecipe(day.lunch),
                        dinner = recipeMap[day.dinner] ?: missingPlanRecipe(day.dinner),
                        snack = recipeMap[day.snack] ?: missingPlanRecipe(day.snack)
                    )
                }
            )
        }

        return DietPlan(
            id = parsed.dietType,
            name = parsed.displayName,
            description = parsed.description,
            benefits = parsed.benefits,
            restrictions = parsed.restrictions,
            weeks = weeks
        )
    }

    private fun PlanRecipeJson.toPlanRecipe(id: String) = PlanRecipe(
        id = id,
        name = name,
        description = description,
        calories = calories,
        protein = protein,
        carbs = carbs,
        fat = fat,
        fiber = fiber,
        prepTime = prepTime,
        cookTime = cookTime,
        ingredients = ingredients,
        instructions = instructions,
        tags = tags
    )

    private fun missingPlanRecipe(id: String) = PlanRecipe(
        id = id,
        name = "Meal",
        description = "",
        calories = 0,
        protein = 0.0,
        carbs = 0.0,
        fat = 0.0,
        fiber = 0.0,
        prepTime = 0,
        cookTime = 0,
        ingredients = emptyList(),
        instructions = emptyList(),
        tags = emptyList()
    )

    companion object {
        // (dietId, fileName) — matches iOS list, order = display order.
        private val DIET_FILES = listOf(
            "mediterranean" to "mediterranean",
            "keto" to "keto",
            "high_protein" to "high_protein",
            "balanced" to "balanced",
            "low_carb" to "low_carb",
            "paleo" to "paleo",
            "whole30" to "whole30",
            "vegan" to "vegan"
        )
    }
}
